### PR TITLE
[GConstruct] Add Customized Edge split

### PR DESCRIPTION
**Pull request by jalencato**
*Issue #, if available:*

*Description of changes:*

#761 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/unit-tests/gconstruct/test_transform.py [hunk 1]
```diff
@@ -641,6 +644,62 @@ def test_custom_label_processor():
...
+
+    split = clp.data_split(data)
+    assert "train_mask" in split
+    assert "val_mask" in split
+    assert "test_mask" in split
+
+    assert_equal(np.squeeze(np.nonzero(split["train_mask"])), index_in_train_idx)
+    assert_equal(np.squeeze(np.nonzero(split["val_mask"])), index_in_val_idx)
+    assert_equal(np.squeeze(np.nonzero(split["test_mask"])), index_in_test_idx)
+    assert len(split["train_mask"]) == 361
```
Comment (GentleZhu) [inline tests/unit-tests/gconstruct/test_transform.py:672]: how 361 is calculated, I am a bit confusing

File: docs/source/tutorials/own-data.rst [hunk 2]
```diff
@@ -178,6 +178,31 @@ These JSON files only need to list the IDs on its own set. For example, in a nod
     ...
     59
 
+For edge data, users can do similar thing to define the customized edge labels. The configuration file should be the same:
+
+.. code-block:: json
+
+    "labels": [
+        {
+            "label_col": "label",
```
[the remark this answers is not in the document]
Comment (jalencato) [reply docs/source/tutorials/own-data.rst:187]: I think the case here is only for link prediction right? If we are using for link prediction, then we do not need these labels. Labels are only for the classification.

File: tests/unit-tests/gconstruct/test_transform.py (hunk 1 above)
Comment (jalencato) [reply tests/unit-tests/gconstruct/test_transform.py:672]: 19 * 19, that is the total number of edges.

File: python/graphstorm/gconstruct/transform.py [hunk 3]
```diff
@@ -1668,10 +1676,16 @@ def parse_label_ops(confs, is_node):
...
-        assert "node_id_col" in confs, "Custom data split only works for nodes."
-        return [CustomLabelProcessor(label_col, label_col, confs["node_id_col"],
-                                     task_type, train_idx, val_idx, test_idx,
-                                     label_stats_type)]
+        if "node_id_col" in confs:
+            return [CustomLabelProcessor(col_name=label_col, label_name=label_col,
+                                         id_col=confs["node_id_col"],
+                                         task_type=task_type, train_idx=train_idx, val_idx=val_idx,
+                                         test_idx=test_idx, stats_type=label_stats_type)]
+        else:
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:1684]: Let's do more check like:
`elif "source_id_col" in confs and "dest_id_col" in confs`.

There can be cases that none of these keys exist in confs.

File: python/graphstorm/gconstruct/transform.py [hunk 4]
```diff
@@ -1451,9 +1451,17 @@ def __call__(self, data):
         dict of Tensors : it contains the labels as well as training/val/test splits.
         """
         label = data[self.col_name] if self.col_name in data else None
-        assert self._id_col in data, \
-                f"The input data does not have ID column {self._id_col}."
-        res = self.data_split(data[self._id_col])
+        if isinstance(self._id_col, str):
+            assert self._id_col in data, \
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:1455]: Please add some comments. For node labels xxx. For edge labels xxx.

File: tests/unit-tests/gconstruct/test_construct_graph.py [hunk 5]
```diff
@@ -719,6 +719,31 @@ def check_classification_no_split(res):
     res = process_labels(data, ops)
     check_classification(res)
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_construct_graph.py:721]: Can we add a separate test function like check_edge_classification_no_split

File: tests/unit-tests/gconstruct/test_construct_graph.py [hunk 6]
```diff
@@ -836,6 +861,37 @@ def check_regression_no_split(res):
     assert np.sum(res['val_mask']) == 1
     assert np.sum(res['test_mask']) == 1
 
+    # Check link prediction on customed masks
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_construct_graph.py:864]: ditto

Comment (classicsong) [reply tests/unit-tests/gconstruct/test_construct_graph.py:864]: Also can you add a test case for link prediction when multiple edge types has their customer splits.

File: python/graphstorm/gconstruct/transform.py [hunk 7]
```diff
@@ -1387,9 +1387,9 @@ def __init__(self, col_name, label_name, id_col, task_type,
         self._id_col = id_col
         self._col_name = col_name
         self._label_name = label_name
-        self._train_idx = set(train_idx.tolist()) if train_idx is not None else None
-        self._val_idx = set(val_idx.tolist()) if val_idx is not None else None
-        self._test_idx = set(test_idx.tolist()) if test_idx is not None else None
+        self._train_idx = set(train_idx) if train_idx is not None else None
```
[the remark this answers is not in the document]
Comment (jalencato) [reply python/graphstorm/gconstruct/transform.py:1390]: Each pair/id in the train_idx should be different, there is no meaning we set a node for one label twice. Using a set will save effort and time to deal with the data.

File: python/graphstorm/gconstruct/file_io.py [hunk 8]
```diff
@@ -30,6 +30,18 @@
 
 from .utils import HDF5Handle, HDF5Array
 
+
+def default_serializer(obj):
```
[the remark this answers is not in the document]
Comment (jalencato) [reply python/graphstorm/gconstruct/file_io.py:34]: That is for json.dump

File: tests/unit-tests/gconstruct/test_construct_graph.py (hunk 5 above)
Comment (jalencato) [reply tests/unit-tests/gconstruct/test_construct_graph.py:721]: Better not to separate the check_edge_classification outside, because we need check_classification which is inside the test_label function, but I can move the link prediction one outside

File: tests/unit-tests/gconstruct/test_construct_graph.py (hunk 6 above)
Comment (jalencato) [reply tests/unit-tests/gconstruct/test_construct_graph.py]: It looks a little weird to write the tests for multiple edge types here. I have drafted another e2e tests for the link prediction with customized label split.

Comment (jalencato) [reply tests/unit-tests/gconstruct/test_construct_graph.py]: And I have moved the link prediction test out to another function.

File: python/graphstorm/gconstruct/file_io.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jalencato) [reply python/graphstorm/gconstruct/file_io.py]: Delete it as we are using data.tolist() now.

File: tests/unit-tests/gconstruct/test_transform.py (hunk 1 above)
Comment (classicsong) [reply tests/unit-tests/gconstruct/test_transform.py:672]: Can you put some comments here?

**Review by classicsong — APPROVED** [review]

File: python/graphstorm/gconstruct/file_io.py [hunk 9]
```diff
@@ -60,9 +66,11 @@ def write_index_json(data, data_file):
     data_file : str
         The data file where the indices are written to.
     """
+    if isinstance(data, np.ndarray):
```
[the remark this answers is not in the document]
Comment (jalencato) [reply python/graphstorm/gconstruct/file_io.py:69]: That is for test only, our previous test has multiple ones directly writing list.
